include/inexor/vulkan-renderer/wrapper/descriptor_builder.hpp

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <vulkan/vulkan_core.h>
+
+#include <cassert>
+#include <string>
+#include <vector>
+
+namespace inexor::vulkan_renderer::wrapper {
+
+class Device;
+class ResourceDescriptor;
+
+class DescriptorBuilder {
+    const Device &m_device;
+    const std::uint32_t m_swapchain_image_count{};
+
+    std::vector<VkDescriptorSetLayoutBinding> m_layout_bindings{};
+    std::vector<VkWriteDescriptorSet> m_write_sets{};
+
+    std::vector<VkDescriptorBufferInfo> m_descriptor_buffer_infos{};
+    std::vector<VkDescriptorImageInfo> m_descriptor_image_infos{};
```

Comment:
Do these need to be members?

Reply:
Yes, they need to be members because of the following problem:
We store the uniform buffer info like this:

```cpp
VkDescriptorBufferInfo uniform_buffer_info{};
uniform_buffer_info.buffer = uniform_buffer;
uniform_buffer_info.offset = 0;
uniform_buffer_info.range = sizeof(T);
```

And later we need this pointer to `uniform_buffer_info` like this:

```cpp
VkWriteDescriptorSet descriptor_write{};
descriptor_write.sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
descriptor_write.dstSet = nullptr;
descriptor_write.dstBinding = binding;
descriptor_write.dstArrayElement = 0;
descriptor_write.descriptorType = VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER;
descriptor_write.descriptorCount = 1;
descriptor_write.pBufferInfo = &m_descriptor_buffer_infos.back();
```

If we leave the current scope, `uniform_buffer_info` will be **invalid** but `descriptor_write` will have the pointer to it.
This is because Vulkan is a C-style API was was not really made for a descriptor pattern in that case I guess?
So we need to store `uniform_buffer_info` in some way until `build` is called. `descriptor_write` will hold a pointer to this member.

This was the reason `VkDescriptorSet` and `VkDescriptorSetLayout` was previously tied to the uniform buffer wrapper.